### PR TITLE
Add config generator

### PR DESCRIPTION
**Pull request by kubasobon**
## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

**Review by kopiczko — COMMENTED** [review]
I finished in the middle as I undestand there is still refactoring going on.

File: service/generator/generator.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+		- .yaml.template is a template
+		- .yaml.template.patch overrides template
+
+		Folder structure:
+			default/
+			    config.yaml
+			    apps/
+			        aws-operator/
+				        ...
+				    azure-operator/
```
Comment (JosephSalisbury) [inline service/generator/generator.go:30]: is the indentation right here?

File: service/generator/generator.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
+package generator
```
Comment (kopiczko) [inline service/generator/generator.go:1]: I think we want to move it to `/pkg` as it will be also used in `/cmd`.

File: service/generator/generator.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+	// installation app-level config
+	configmapTemplatePatchFile = "configmap-values.yaml.template.patch"
+)
+
+type Filesystem interface {
+	Exists(string) bool
+	ReadFile(string) ([]byte, error)
+	ReadDir(string) ([]os.FileInfo, error)
+}
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply service/generator/generator.go]: Moved.

File: service/generator/generator.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+type Config struct {
+	Fs Filesystem
+}
+
+type Generator struct {
+	fs Filesystem
+}
+
+func New(config *Config) (*Generator, error) {
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply service/generator/generator.go]: Check added.

File: service/generator/generator.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+// 1. Get configmap context and patch it with installation-specific overrides (if
+//    available)
+// 2. Get app-specific configmap template and patch it with installation-specific
+//    app overrides (if available)
+// 3. Render app configmap template (result of 2.) with context values (result
+//    of 1.)
+// 4. Get secrets context
+// 5. Get app-specific secrets template.
+// 6. Render app secrets template (result of 4.) with installation secrets (result of 5.)
+func (g Generator) GenerateConfig(installation, app string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply service/generator/generator.go]: Named.

File: service/generator/generator.go (hunk 1 above)
Comment (kubasobon) [reply service/generator/generator.go]: Re-tabbed.

File: service/generator/generator.go (hunk 2 above)
Comment (kubasobon) [reply service/generator/generator.go]: Moved to `/pkg`.

File: service/generator/generator_test.go [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+	"os"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/giantswarm/microerror"
+)
+
+func TestGenerator_GenerateConfig(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply service/generator/generator_test.go]: I agree. I'll migrate the test cases to golden.

**Comment by kubasobon** [comment]
@kopiczko PTAL now ;)

**Review by kopiczko — COMMENTED** [review]
I'll check the rest after meetings

File: pkg/generator/generator.go [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+	)
+	if err != nil {
+		return "", "", microerror.Mask(err)
+	}
+
+	// 7.
+	secrets, err = g.renderTemplate(secretsTemplate, secretsContext)
+	if err != nil {
+		return "", "", microerror.Mask(err)
+	}
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply pkg/generator/generator.go]: `secrets.yaml` exists per installation and has no patch, so it's different from 3.

File: pkg/generator/generator.go [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+	out := bytes.NewBuffer([]byte{})
+	err = t.Execute(out, c)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return out.String(), nil
+}
+
+func (g Generator) addIncludeFilesToTemplate(t *template.Template) error {
```
Comment (kopiczko) [inline pkg/generator/generator.go:270]: Correct me if I'm wrong but it feels unnecessary. I think include could find template file dynamically as it's called. I noticed that because I wasn't sure why we add `sprig.FuncMap` in so many places.

**Review by kopiczko — COMMENTED** [review]
Next batch: still need to check tests.

File: pkg/generator/generator.go [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+	result, err := g.renderTemplate(string(templateBytes), context)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return result, nil
+}
+
+func applyPatch(base, patch []byte) (string, error) {
+	var basePathSvc *pathmodifier.Service
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply pkg/generator/generator.go:196]: The issues with non-string types are solved for now. I couldn't find a corner-case where they are broken, so until then let's not rip out and replace merging logic: [test case 3](https://github.com/giantswarm/config-controller/pull/12/files#diff-8c3659b90d3a1edd973b3fe009c1bae9621a7676d1e9262c84b14c96667a6e6c)

[the remark this answers is not in the document]
Comment (kopiczko) [reply pkg/generator/generator.go:196]: Besides having `.` in the key name. But I thing we won't hit that.

File: pkg/generator/testdata/case6.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  - large
+---
+path: configmap-values.yaml.golden
+data: |
+  answer: 42
+  availableInstances:
+  - small
+  - medium
+  - large
+  region: us-east-1
```
Comment (kopiczko) [inline pkg/generator/testdata/case6.yaml:31]: Those tests are nice!

**Review by kopiczko — APPROVED** [review]

File: pkg/generator/generator.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (kubasobon) [reply pkg/generator/generator.go:194]: I think only secrets have paths with a `.` ([example](https://github.com/giantswarm/aws-operator/blob/master/helm/aws-operator/templates/secret.yaml#L10-L11)), so as long as we don't want to patch secrets, we are good.

File: pkg/generator/generator.go [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+	g := Generator{
+		fs: config.Fs,
+	}
+
+	return &g, nil
+}
+
+// GenerateConfig creates final configmap values and secret values for helm to
+// use by performing the following operations:
+// 1. Get configmap context and patch it with installation-specific overrides (if
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply pkg/generator/generator.go]: Replaced. I spent too much time with Jinja2, where it's context :P

File: pkg/generator/generator.go [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+	}
+
+	funcMap := sprig.FuncMap()
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		baseName := path.Base(file.Name())
+		if strings.ContainsRune(baseName, '.') {
```
[the remark this answers is not in the document]
Comment (kubasobon) [reply pkg/generator/generator.go]: Gone with the function.

File: pkg/generator/generator.go (hunk 8 above)
Comment (kubasobon) [reply pkg/generator/generator.go]: Cleaned up `include` logic.
